lib/client-retry.js

```diff
@@ -0,0 +1,151 @@
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
+                    const waitTime = this.retryWaitTime(e.retryAfter);
+                    console.log(colors.red(`Waiting for ${waitTime}ms and trying again`));
+                    await sleep(waitTime);
+                } else {
+                    non429Response = true;
+                    error = e;
+                }
+            }
+        }
+        if (!non429Response) {
+            throw new Error(`Running into 429s after ${retryCount} attempts. Will not continue retrying. Try again in a few minutes.`);
+
+        }
+        // pass through error
+        if (error) {
+            throw error;
+        }
+        return response;
+    }
+
+    // retry on 429s
+    async unregister() {
+        let response, error;
+        let non429Response = false;
+        let retryCount = 0;
+        while (!non429Response && retryCount <= this.max429RetryCount) {
+            retryCount++;
+            try {
+                response = await super.unregister();
+                non429Response = true;
+            } catch (e) {
+                // string comparision instead of `instanceof` since `instanceof` causes issues
+                if (e.code === TOO_MANY_REQUEST_ERROR_CODE && e.name === TOO_MANY_REQUEST_ERROR) {
+                    non429Response = false;
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
+                    const waitTime = this.retryWaitTime(e.retryAfter);
+                    console.log(colors.red(`Waiting for ${waitTime}ms and trying again`));
+                    await sleep(waitTime);
+                } else {
+                    non429Response = true;
+                    error = e;
+                }
+            }
+        }
+
+        if (!non429Response) {
+            throw new Error(`Running into 429s after ${retryCount} attempts. Will not continue retrying. Try again in a few minutes.`);
+
+        }
+        // pass through error
+        if (error) {
+            throw error;
+        }
+        return response;
+    }
+    async process(source, renditions, userData) {
+        let response, error;
+        let non429Response = false;
+        let retryCount = 0;
+        while (!non429Response && retryCount <= this.max429RetryCount) {
+            retryCount++;
+            try {
+                response = await super.process(source, renditions, userData);
+                console.log('Successful call to /process', retryCount);
+                non429Response = true;
+            } catch (e) {
+                // string comparision instead of `instanceof` since `instanceof` causes issues
+                if (e.code === TOO_MANY_REQUEST_ERROR_CODE && e.name === TOO_MANY_REQUEST_ERROR) {
+                    non429Response = false;
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
+                    const waitTime = this.retryWaitTime(e.retryAfter);
```

Comment:
can `e.retryAfter` be undefined? good idea to add a test.

Reply:
I have a test at the micro level (asset-compute.js): https://github.com/adobe/asset-compute-client/blob/NUI-1650/test/assetcompute.test.js#L184-L208
I can add one at the client-retry.js level too